### PR TITLE
Implement AllowIcmp() for AWS and GCP provider.

### PR DESCRIPTION
**Pull request by dPhanekham**
I found that the ping benchmark with --ping_also_run_using_external_ip=True would fail when trying to perform tests with external IP addresses.

The benchmark makes a call to AllowIcmp(), which is not implemented for multiple cloud providers. So I have implemented that function for Google Cloud and now ping tests using external IPs function as intended.

external ping tests are also failing on AWS and I'm working on a fix for those

### CODE REVIEW

**Comment by dPhanekham** [comment]
Also added a fix for aws ping tests

File: perfkitbenchmarker/providers/aws/aws_network.py [hunk 1]
```diff
@@ -64,6 +64,28 @@ def __init__(self):
     self.firewall_set = set()
     self._lock = threading.Lock()
 
+  def AllowIcmp(self, vm):
+    """Opens the ICMP protocol on the firewall.
+    Args:
+      vm: The BaseVirtualMachine object to open the ICMP protocol for.
+    """
+    print("ALLOWING ICMP")
```
Comment (miketruty) [inline perfkitbenchmarker/providers/aws/aws_network.py:72]: Why do you print status here and not in gce_network.py?

Comment (dPhanekham) [reply perfkitbenchmarker/providers/aws/aws_network.py]: Forgot to remove a debug statement. It is gone now

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py [hunk 2]
```diff
@@ -103,7 +103,7 @@ def _RunPing(sending_vm, receiving_vm, receiving_ip, ip_type):
   Returns:
     A list of samples, with one sample for each metric.
   """
-  if not sending_vm.IsReachable(receiving_vm):
+  if not sending_vm.IsReachable(receiving_vm) and ip_type == 'internal':
```
Comment (miketruty) [inline perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:106]: Since IsReachable() is only valid on ip_type internal, consider inverting order of clauses?

Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py]: changed

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py [hunk 3]
```diff
@@ -103,7 +103,7 @@ def _RunPing(sending_vm, receiving_vm, receiving_ip, ip_type):
   Returns:
     A list of samples, with one sample for each metric.
   """
-  if not sending_vm.IsReachable(receiving_vm):
+  if ip_type == 'internal' and not sending_vm.IsReachable(receiving_vm):
```
Comment (dlott) [inline perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:106]: Why isn't this cased the same as `IpAddressSubset.INTERNAL`

File: perfkitbenchmarker/providers/aws/aws_network.py [hunk 4]
```diff
@@ -64,6 +64,27 @@ def __init__(self):
     self.firewall_set = set()
     self._lock = threading.Lock()
 
+  def AllowIcmp(self, vm):
+    """Opens the ICMP protocol on the firewall.
+    Args:
+      vm: The BaseVirtualMachine object to open the ICMP protocol for.
+    """
+    entry = (-1, -1, vm.region, vm.group_id)
```
Comment (dlott) [inline perfkitbenchmarker/providers/aws/aws_network.py:72]: Other `firewall_set` entries include cidr as the fifth tuple.  Should we have different firewall_set for tcp/udp and icmp?

File: perfkitbenchmarker/providers/gcp/gce_network.py [hunk 5]
```diff
@@ -586,6 +586,30 @@ def DisallowAllPorts(self):
...
+      vm: The BaseVirtualMachine object to open the ICMP protocol for.
+    """
+    if vm.is_static:
+      return
+    with self._lock:
+      firewall_name = ('perfkit-firewall-%s-%d-%d' %
+                       (FLAGS.run_uri, -1, -1))
+
+      key = (vm.project, -1, -1)
+      if key in self.firewall_rules:
```
Comment (dlott) [inline perfkitbenchmarker/providers/gcp/gce_network.py:602]: similar question about comingling tcp/udp and icmp rules in the same dictionary.

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py (hunk 3 above)
Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:106]: It's just a string passed by the Run() function directly above it, but I can change it to INTERNAL/EXTERNAL if that is more consistent

File: perfkitbenchmarker/providers/aws/aws_network.py (hunk 4 above)
Comment (dPhanekham) [reply perfkitbenchmarker/providers/aws/aws_network.py]: Changed to add cidr as a fifth tuple. I don't think we need a separate firewall set for ICMP, they are all firewall rules and with the way the entries are formatted, there shouldn't be duplicates

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py (hunk 3 above)
Comment (dlott) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:106]: Hardcoding the same constant in two different places is not as good as using the same shared constant.  You could define a new one here, but if there's already a enum that does what you want, seems better to use it.

File: perfkitbenchmarker/providers/aws/aws_network.py (hunk 4 above)
Comment (dlott) [reply perfkitbenchmarker/providers/aws/aws_network.py]: They both (int, int, string, string, string) Seems like there could be duplicates.  Please either use separate sets or add the protocol to the tuple, or add tests to prevent future maintainers from breaking the implicit formatting assumptions.

Comment (dPhanekham) [reply perfkitbenchmarker/providers/aws/aws_network.py]: I see where the overlap could be now. I've moved icmp firewall rules to their own sets for both aws and gcp

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py (hunk 3 above)
Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py]: Changed to use existing enum

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py [hunk 6]
```diff
@@ -103,11 +107,11 @@ def _RunPing(sending_vm, receiving_vm, receiving_ip, ip_type):
   Returns:
     A list of samples, with one sample for each metric.
   """
-  if not sending_vm.IsReachable(receiving_vm):
+  if ip_type == vm_util.IpAddressSubset.INTERNAL and not sending_vm.IsReachable(receiving_vm):
```
Comment (dlott) [inline perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:110]: Please update the method documentation.

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py [hunk 7]
```diff
@@ -103,11 +107,11 @@ def _RunPing(sending_vm, receiving_vm, receiving_ip, ip_type):
...
     logging.warn('%s is not reachable from %s', receiving_vm, sending_vm)
     return []
 
-  logging.info('Ping results (ip_type = %s):', ip_type)
+  logging.info('Ping results (ip_type = %s):' % ip_type)
```
Comment (dlott) [inline perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:114]: Please revert this change (we have a different internal linter that complains about `logging-not-lazy`)  We're working to align both of them...

File: perfkitbenchmarker/providers/aws/aws_network.py [hunk 8]
```diff
@@ -62,8 +62,32 @@ class AwsFirewall(network.BaseFirewall):
...
+    self.firewall_icmp_set = set()
     self._lock = threading.Lock()
 
+  def AllowIcmp(self, vm):
+    """Opens the ICMP protocol on the firewall.
+    Args:
+      vm: The BaseVirtualMachine object to open the ICMP protocol for.
+    """
+    source = '0.0.0.0/0'
+    # type, code, region, group_id, cidr
```
Comment (dlott) [inline perfkitbenchmarker/providers/aws/aws_network.py:74]: This doesn't match the fields of the tuple?

Comment (dPhanekham) [reply perfkitbenchmarker/providers/aws/aws_network.py:74]: In the aws cli, when specifying an ICMP firewall rule, the from-port and to-port numbers actually specify the type and code of ICMP packets allowed, with -1 meaning all types/codes.

https://docs.aws.amazon.com/cli/latest/reference/ec2/authorize-security-group-ingress.html
https://www.iana.org/assignments/icmp-parameters/icmp-parameters.xhtml

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py (hunk 6 above)
Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:113]: updated the documentation to reflect the enum being used for ip_type

File: perfkitbenchmarker/providers/aws/aws_network.py (hunk 8 above)
Comment (dlott) [reply perfkitbenchmarker/providers/aws/aws_network.py:74]: Ahhh. Both the API is confusing (can you include a note that port=type and link to the CLI) and you've got `code` in the tuple, but not in `authorize_cmd`.  Since we've got separate sets now, you should just be able to drop it from the tuple, right?

File: perfkitbenchmarker/providers/gcp/gce_network.py [hunk 9]
```diff
@@ -585,6 +586,30 @@ def DisallowAllPorts(self):
...
+
+    Args:
+      vm: The BaseVirtualMachine object to open the ICMP protocol for.
+    """
+    if vm.is_static:
+      return
+    with self._lock:
+      firewall_name = ('perfkit-firewall-icmp-%s-%d-%d' %
+                       (FLAGS.run_uri, -1, -1))
+      key = (vm.project, -1, -1)
```
Comment (dlott) [inline perfkitbenchmarker/providers/gcp/gce_network.py:603]: There's no specifying of ports/types/codes here?  Can we just remove the -1 from the key and the name?

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py (hunk 7 above)
Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py]: change reverted

File: perfkitbenchmarker/providers/aws/aws_network.py (hunk 8 above)
Comment (dPhanekham) [reply perfkitbenchmarker/providers/aws/aws_network.py]: I've inserted a comment explaining the port=type connection and a included a link. 

I was going to leave the type and code in the tuple in case anyone wanted to add the ability to specify the type and code in the future, but I've removed it now for the sake of simplicity.

File: perfkitbenchmarker/providers/gcp/gce_network.py (hunk 9 above)
Comment (dPhanekham) [reply perfkitbenchmarker/providers/gcp/gce_network.py]: Yes. The icmp firewall stuff in the gcloud cli doesn't have the same quirks as the aws one, so I've removed the ports from the key. 

I also added the vm cidr to the key, similar to the AllowPort() method so it doesn't have any collisions when there are multiple networks

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py [hunk 10]
```diff
@@ -22,6 +22,7 @@
 from perfkitbenchmarker import configs
 from perfkitbenchmarker import flags
 from perfkitbenchmarker import sample
+from perfkitbenchmarker import vm_util
 import re
 
 flags.DEFINE_boolean('ping_also_run_using_external_ip', False,
```
Comment (dlott) [inline perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:28]: Per conversation, please use the ShouldRunOnExternalIpAddress() style.

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py [hunk 11]
```diff
@@ -113,7 +120,13 @@ def _RunPing(sending_vm, receiving_vm, receiving_ip, ip_type):
   stats = re.findall('([0-9]*\\.[0-9]*)', stdout.splitlines()[-1])
   assert len(stats) == len(METRICS), stats
   results = []
-  metadata = {'ip_type': ip_type,
+
+  if ip_type == vm_util.IpAddressSubset.INTERNAL:
+    ip_string = 'internal'
```
Comment (dlott) [inline perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:125]: Can you unify this with the constants in netperf_benchmark?  Defining them in vm_util seems reasonable for now.

File: perfkitbenchmarker/providers/gcp/gce_network.py [hunk 12]
```diff
@@ -585,6 +586,37 @@ def DisallowAllPorts(self):
...
+      vm: The BaseVirtualMachine object to open the ICMP protocol for.
+    """
+    if vm.is_static:
+      return
+    with self._lock:
+      if vm.cidr:  # Allow multiple networks per zone.
+        cidr_string = network.BaseNetwork.FormatCidrString(vm.cidr)
+        firewall_name = ('perfkit-firewall-icmp-%s-%s' %
+                         (cidr_string, FLAGS.run_uri))
+        key = (vm.project, vm.cidr)
```
Comment (cwilkes) [inline perfkitbenchmarker/providers/gcp/gce_network.py:605]: It looks like GCEFirewall doesn't use vm.cidr but vm.network.network_resource.name when generating a rule.
I take it there's a 1:1 mapping between cidr and the network name?

Comment (dPhanekham) [reply perfkitbenchmarker/providers/gcp/gce_network.py:605]: Yes, I tried to implement this in the same way as in the AllowPort method.

If the cidr string is 10-0-1-0-24, then the network resource name would be something like
pkb-network-multi-10-0-1-0-24-(run-uri).

I have also tested and confirmed that this works with all of the network configs in gcp_network_test

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py (hunk 11 above)
Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py]: I've made a new enum in vm_util and changed netperf,iperf,ping to use it

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py (hunk 10 above)
Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py]: Changed to use the style we discussed

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py [hunk 13]
```diff
@@ -25,10 +25,6 @@
 from perfkitbenchmarker import vm_util
 import re
 
-flags.DEFINE_boolean('ping_also_run_using_external_ip', False,
```
Comment (dlott) [inline perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:27]: Looks like @miketruty 's new tutorials used this flag.  
https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/d486aa9ba587c44dc3edc43f75b8fa0d7222f1f7/tutorials/inter_region_reports/data/simple_region_latency.yaml#L16

Are they working currently given that we don't have the firewalls in place?  
Preferences on updating these all together or seperately?

Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py:27]: There is currently an error when using external IPs in the tutorial, which is what originally prompted this PR. I'll also update the tutorial to use the new stuff

File: perfkitbenchmarker/linux_benchmarks/ping_benchmark.py (hunk 10 above)
Comment (dlott) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py]: Please update the breaking changes section with the deleted flag and new usage.  Otherwise we'll let @cwilkes take one last look.

Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/ping_benchmark.py]: I've added an entry in the breaking changes section
